src/compiler/checker.ts

```diff
@@ -11719,6 +11719,11 @@ namespace ts {
                 forEachMappedTypePropertyKeyTypeAndIndexSignatureKeyType(modifiersType, include, keyofStringsOnly, addMemberForKeyType);
             }
             else {
+                forEachType(constraintType, t => {

```

Comment:
Hmmmm, so you're doing this in a 2nd pass because the `keyof` part's lower bound becomes `string` which then destroys the concrete property when it's union'd. Rather than having this 2-pass thing, I'd rather just edit `getLowerBoundOfKeyType` to pass `noReductions` to `mapType`.

Reply:
@weswigham ok, I didn't know much about `noReductions` 😅 it seems to work well - I've just pushed out the requested change